Add particle generators namespace

gassmoeller opened the pull request:
A typical operation using the particles is to generate them in the first place, which is currently a tedious work the user has to do. This PR moves over some code from geodynamics/aspect (https://github.com/geodynamics/aspect/blob/master/source/particle/generator/reference_cell.cc) that generates particles at user-defined locations on the reference cell in each cell of the given triangulation. These locations could be cell centers, arbitrary locations, or patterns such as quadrature points / finite-element support points. I modelled the namespace and function interface after the GridGenerator namespace, but let me know if you prefer a different interface.

This will hopefully be useful for @blaisb and @aggarwal-ashna.

bangerth commented on include/deal.II/particles/particle_generator.h:
```diff
@@ -0,0 +1,68 @@
+   * A namespace that contains all classes that are related to the particle
+   * generation.
+   */
+  namespace Generator
```
We have not designed namespace names well :-( There is `Particles` (plural) and `Generator` (singular) just like there is `GridGenerator` and `VectorTools`. I don't know what to do about this, but wished we had generally used plural form...

bangerth commented on include/deal.II/particles/particle_generator.h:
```diff
@@ -0,0 +1,68 @@
+     * Particles will be generated in every cell at these locations.
+     *
+     * @param particle_handler The particle handler that will take ownership
+     *  of the generated particles.
```
You don't indent the second line above or below, so maybe not here either.
```suggestion
     * of the generated particles.
```

bangerth commented on include/deal.II/particles/particle_generator.h:
```diff
@@ -0,0 +1,68 @@
+      const parallel::distributed::Triangulation<dim, spacedim> &triangulation,
+      const std::vector<Point<dim>> & particle_reference_locations,
+      ParticleHandler<dim, spacedim> &particle_handler,
+      const Mapping<dim, spacedim> &  mapping = MappingQ1<dim, spacedim>());
```
Don't we have `StaticMappingQ1::mapping` somewhere for this purpose?

bangerth commented on include/deal.II/particles/particle_generator.h:
```diff
@@ -0,0 +1,68 @@
+    template <int dim, int spacedim = dim>
+    void
+    regular_reference_locations(
+      const parallel::distributed::Triangulation<dim, spacedim> &triangulation,
```
I think you can just use a regular `::Triangulation` here.

bangerth commented on include/deal.II/particles/particle_generator.h:
```diff
@@ -0,0 +1,68 @@
+// ---------------------------------------------------------------------
```
Why not call the file `generator.h`?

bangerth commented on source/particles/particle_generator.cc:
```diff
@@ -0,0 +1,84 @@
+      ParticleHandler<dim, spacedim> &particle_handler,
+      const Mapping<dim, spacedim> &  mapping)
+    {
+      types::particle_index n_particles_to_generate =
```
```suggestion
      const types::particle_index n_particles_to_generate =
```

bangerth commented on source/particles/particle_generator.cc:
```diff
@@ -0,0 +1,84 @@
+      // The local start index is the number of all particles
+      // generated on lower MPI ranks.
+      types::particle_index local_start_index = 0;
+      MPI_Scan(&n_particles_to_generate,
```
@kronbichler has taught me in one of his previous patches that `MPI_Exscan` is the more modern variant. Can you explore this?

bangerth commented on source/particles/particle_generator.cc:
```diff
@@ -0,0 +1,84 @@
+               DEAL_II_PARTICLE_INDEX_MPI_TYPE,
+               MPI_SUM,
+               triangulation.get_communicator());
+      local_start_index -= n_particles_to_generate;
```
If I recall correctly, the exscan avoids this operation.

gassmoeller commented:
I think I addressed all your comments. At the moment it is still simple to rename files and namespaces. Strangely I now get different test results for debug and release mode (though only numbers that are very close to zero).

bangerth commented on include/deal.II/particles/generators.h:
```diff
@@ -0,0 +1,69 @@
+  {
+    /**
+     * A function that generates particles in every cell at specified @p particle_reference_locations.
+     * The total number of particles that is added to the @p partiche_handler object is
```
```suggestion
     * The total number of particles that is added to the @p particle_handler object is
```

gassmoeller commented:
Fixed the typo.

bangerth commented:
/retest

masterleinad commented:
@bangerth I guess you meant
/rebuild

bangerth commented:
Ah, there are a bunch of failures:
```
/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:26:39: error: template parameter redefines default argument

    template <int dim, int spacedim = dim>

                                      ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/include/deal.II/particles/generators.h:55:39: note: previous default template argument defined here

    template <int dim, int spacedim = dim>

                                      ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:49:22: error: use of undeclared identifier 'DEAL_II_PARTICLE_INDEX_MPI_TYPE'

                     DEAL_II_PARTICLE_INDEX_MPI_TYPE,

                     ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:68:35: error: implicit instantiation of undefined template 'dealii::Particles::ParticleHandler<1, 1>'

                  particle_handler.insert_particle(particle, cell);

                                  ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/build/source/particles/generators.inst:16:2: note: in instantiation of function template specialization 'dealii::Particles::Generators::regular_reference_locations<1, 1>' requested here

 regular_reference_locations< 1 ,  1 >(

 ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/include/deal.II/particles/particle_iterator.h:28:9: note: template is declared here

  class ParticleHandler;

        ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:74:23: error: implicit instantiation of undefined template 'dealii::Particles::ParticleHandler<1, 1>'

      particle_handler.update_cached_numbers();

                      ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/include/deal.II/particles/particle_iterator.h:28:9: note: template is declared here

  class ParticleHandler;

        ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:68:35: error: implicit instantiation of undefined template 'dealii::Particles::ParticleHandler<1, 2>'

                  particle_handler.insert_particle(particle, cell);

                                  ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/build/source/particles/generators.inst:36:2: note: in instantiation of function template specialization 'dealii::Particles::Generators::regular_reference_locations<1, 2>' requested here

 regular_reference_locations< 1 ,  2 >(

 ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/include/deal.II/particles/particle_iterator.h:28:9: note: template is declared here

  class ParticleHandler;

        ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:74:23: error: implicit instantiation of undefined template 'dealii::Particles::ParticleHandler<1, 2>'

      particle_handler.update_cached_numbers();

                      ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/include/deal.II/particles/particle_iterator.h:28:9: note: template is declared here

  class ParticleHandler;

        ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:68:35: error: implicit instantiation of undefined template 'dealii::Particles::ParticleHandler<1, 3>'

                  particle_handler.insert_particle(particle, cell);

                                  ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/build/source/particles/generators.inst:56:2: note: in instantiation of function template specialization 'dealii::Particles::Generators::regular_reference_locations<1, 3>' requested here

 regular_reference_locations< 1 ,  3 >(

 ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/include/deal.II/particles/particle_iterator.h:28:9: note: template is declared here

  class ParticleHandler;

        ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:74:23: error: implicit instantiation of undefined template 'dealii::Particles::ParticleHandler<1, 3>'

      particle_handler.update_cached_numbers();

                      ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/include/deal.II/particles/particle_iterator.h:28:9: note: template is declared here

  class ParticleHandler;

        ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:68:35: error: implicit instantiation of undefined template 'dealii::Particles::ParticleHandler<2, 2>'

                  particle_handler.insert_particle(particle, cell);

                                  ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/build/source/particles/generators.inst:96:2: note: in instantiation of function template specialization 'dealii::Particles::Generators::regular_reference_locations<2, 2>' requested here

 regular_reference_locations< 2 ,  2 >(

 ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/include/deal.II/particles/particle_iterator.h:28:9: note: template is declared here

  class ParticleHandler;

        ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:74:23: error: implicit instantiation of undefined template 'dealii::Particles::ParticleHandler<2, 2>'

      particle_handler.update_cached_numbers();

                      ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/include/deal.II/particles/particle_iterator.h:28:9: note: template is declared here

  class ParticleHandler;

        ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:68:35: error: implicit instantiation of undefined template 'dealii::Particles::ParticleHandler<2, 3>'

                  particle_handler.insert_particle(particle, cell);

                                  ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/build/source/particles/generators.inst:116:2: note: in instantiation of function template specialization 'dealii::Particles::Generators::regular_reference_locations<2, 3>' requested here

 regular_reference_locations< 2 ,  3 >(

 ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/include/deal.II/particles/particle_iterator.h:28:9: note: template is declared here

  class ParticleHandler;

        ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:74:23: error: implicit instantiation of undefined template 'dealii::Particles::ParticleHandler<2, 3>'

      particle_handler.update_cached_numbers();

                      ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/include/deal.II/particles/particle_iterator.h:28:9: note: template is declared here

  class ParticleHandler;

        ^

[ 85%] Built target obj_differentiation_ad_debug

[ 86%] Building CXX object source/numerics/CMakeFiles/obj_numerics_debug.dir/vector_tools_project_qpmf.cc.o

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:68:35: error: implicit instantiation of undefined template 'dealii::Particles::ParticleHandler<3, 3>'

                  particle_handler.insert_particle(particle, cell);

                                  ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/build/source/particles/generators.inst:176:2: note: in instantiation of function template specialization 'dealii::Particles::Generators::regular_reference_locations<3, 3>' requested here

 regular_reference_locations< 3 ,  3 >(

 ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/include/deal.II/particles/particle_iterator.h:28:9: note: template is declared here

  class ParticleHandler;

        ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/source/particles/generators.cc:74:23: error: implicit instantiation of undefined template 'dealii::Particles::ParticleHandler<3, 3>'

      particle_handler.update_cached_numbers();

                      ^

/Users/jenkins/jenkins/workspace/dealii-OSX_PR-8577/include/deal.II/particles/particle_iterator.h:28:9: note: template is declared here

  class ParticleHandler;

        ^
```

masterleinad commented:
```
/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:16:2: error: variable has incomplete type 'void' [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:16:29: error: expected ';' at end of declaration [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:16:29: error: expected unqualified-id [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:16:29: error: variable cannot be defined in an explicit instantiation; if this declaration is meant to be a variable definition, remove the 'template' keyword [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:36:2: error: variable has incomplete type 'void' [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:36:29: error: expected ';' at end of declaration [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:36:29: error: expected unqualified-id [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:36:29: error: variable cannot be defined in an explicit instantiation; if this declaration is meant to be a variable definition, remove the 'template' keyword [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:56:2: error: variable has incomplete type 'void' [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:56:29: error: expected ';' at end of declaration [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:56:29: error: expected unqualified-id [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:56:29: error: variable cannot be defined in an explicit instantiation; if this declaration is meant to be a variable definition, remove the 'template' keyword [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:96:2: error: variable has incomplete type 'void' [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:96:29: error: expected ';' at end of declaration [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:96:29: error: expected unqualified-id [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:96:29: error: variable cannot be defined in an explicit instantiation; if this declaration is meant to be a variable definition, remove the 'template' keyword [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:116:2: error: variable has incomplete type 'void' [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:116:29: error: expected ';' at end of declaration [clang-diagnostic-error]

/jenkins/workspace/dealii-tidy_PR-8577/build/source/particles/generators.inst:116:29: error: variable cannot be defined in an explicit instantiation; if this declaration is meant to be a variable definition, remove the 'template' keyword 
```

gassmoeller commented:
The problems of many possible configurations :smile:. I think I fixed it.

masterleinad commented on doc/news/changes/minor/20190814ReneGassmoeller:
```diff
@@ -0,0 +1,4 @@
+Added: Add a Particles::Generator namespace and a first implementation that
```
```suggestion
Added: Add a Particles::Generators namespace and a first implementation that
```
But we can also go with a follow-up pull request to avoid rerunning the tester.
